File: ranking_engine.js
/**
 * JavaScript Ranking Engine - Performance comparison with Python version
 */

class Post {
    constructor(postId, options = {}) {
        this.postId = postId;
        this.likes = options.likes || 0;
        this.comments = options.comments || 0;
        this.shares = options.shares || 0;
        this.upvotes = options.upvotes || 0;
        this.downvotes = options.downvotes || 0;
        this.timestamp = options.timestamp || Date.now() / 1000;
        this.baseScore = this.likes + this.comments * 2 + this.shares * 3;
    }

    updateBaseScore() {
        this.baseScore = this.likes + this.comments * 2 + this.shares * 3;
    }
}

class StreamingRankingEngine {
    constructor(batchSize = 100000, maxHeapSize = 1000) {
        this.batchSize = batchSize;
        this.maxHeapSize = maxHeapSize;
    }

    hotScore(post) {
        const order = Math.log10(Math.max(Math.abs(post.upvotes - post.downvotes), 1));
        const sign = post.upvotes - post.downvotes > 0 ? 1 : 
                    post.upvotes - post.downvotes < 0 ? -1 : 0;
        const seconds = Date.now() / 1000 - post.timestamp - 1134028003;
        return Math.round((sign * order + seconds / 45000) * 10000000) / 10000000;
    }

    engagementScore(post) {
        const totalEngagement = post.likes + post.comments * 2 + post.shares * 3;
        const timeFactor = Date.now() / 1000 - post.timestamp + 1;
        return totalEngagement / timeFactor;
    }

    timeDecayScore(post, decayRate = 0.01) {
        const timeElapsed = Date.now() / 1000 - post.timestamp;
        return post.baseScore * Math.exp(-decayRate * timeElapsed);
    }

    calculateScore(post, algorithm, options = {}) {
        switch (algorithm) {
            case "hot_score":
                return this.hotScore(post);
            case "engagement_score":
                return this.engagementScore(post);
            case "time_decay":
                const decayRate = options.decayRate || 0.01;
                return this.timeDecayScore(post, decayRate);
            default:
                throw new Error(`Unknown algorithm: ${algorithm}`);
        }
    }

    rankPostsStreaming(postIterator, algorithm = "hot_score", topK = 100, options = {}) {
        const startTime = performance.now();
        const startMemory = this.getMemoryUsage();

        const minHeap = [];
        let totalPosts = 0;
        let batchCount = 0;

        try {
            while (true) {
                const batch = [];
                for (let i = 0; i < this.batchSize; i++) {
                    const next = postIterator.next();
                    if (next.done) break;
                    batch.push(next.value);
                    totalPosts++;
                }

                if (batch.length === 0) break;

                batchCount++;
                console.log(`Processing batch ${batchCount} (${batch.length} posts, total: ${totalPosts.toLocaleString()})`);

                for (const post of batch) {
                    const score = this.calculateScore(post, algorithm, options);

                    if (minHeap.length < topK) {
                        minHeap.push([score, post.postId, post]);
                        this.heapifyUp(minHeap, minHeap.length - 1);
                    } else {
                        if (score > minHeap[0][0]) {
                            minHeap[0] = [score, post.postId, post];
                            this.heapifyDown(minHeap, 0);
                        }
                    }
                }

                if (batchCount % 10 === 0) {
                    const currentMemory = this.getMemoryUsage();
                    console.log(`  Memory usage: ${currentMemory.toFixed(1)} MB`);
                }
            }
        } catch (error) {
            console.error(`Error during streaming ranking: ${error}`);
            throw error;
        }

        const topPosts = minHeap
            .sort((a, b) => b[0] - a[0])
            .map(([score, postId, post]) => [score, post]);

        const endTime = performance.now();
        const endMemory = this.getMemoryUsage();

        return {
            topPosts: topPosts,
            totalPostsProcessed: totalPosts,
            memoryUsageMb: endMemory,
            processingTimeSeconds: (endTime - startTime) / 1000,
            algorithm: algorithm,
            batchSize: this.batchSize
        };
    }

    heapifyUp(heap, index) {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (heap[parentIndex][0] <= heap[index][0]) break;
            [heap[parentIndex], heap[index]] = [heap[index], heap[parentIndex]];
            index = parentIndex;
        }
    }

    heapifyDown(heap, index) {
        while (true) {
            let smallest = index;
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;

            if (leftChild < heap.length && heap[leftChild][0] < heap[smallest][0]) {
                smallest = leftChild;
            }
            if (rightChild < heap.length && heap[rightChild][0] < heap[smallest][0]) {
                smallest = rightChild;
            }

            if (smallest === index) break;
            [heap[index], heap[smallest]] = [heap[smallest], heap[index]];
            index = smallest;
        }
    }

    getMemoryUsage() {
        if (typeof process !== 'undefined' && process.memoryUsage) {
            return process.memoryUsage().heapUsed / 1024 / 1024;
        }
        return 0;
    }

    rankPostsFromList(posts, algorithm = "hot_score", topK = 100, options = {}) {
        const postIterator = {
            next: function() {
                if (this.index >= posts.length) {
                    return { done: true };
                }
                return { done: false, value: posts[this.index++] };
            },
            index: 0
        };

        return this.rankPostsStreaming(postIterator, algorithm, topK, options);
    }
}

function generateTestPosts(numPosts) {
    const baseTime = Date.now() / 1000;
    const posts = [];

    for (let i = 0; i < numPosts; i++) {
        const likes = (i * 7) % 10000 + 1;
        const comments = (i * 3) % 1000 + 1;
        const shares = (i * 5) % 500 + 1;
        const upvotes = likes + (i % 100);
        const downvotes = (i % 50);
        const timestamp = baseTime - (i % (86400 * 30));

        posts.push(new Post(`post_${i.toString().padStart(8, '0')}`, {
            likes,
            comments,
            shares,
            upvotes,
            downvotes,
            timestamp
        }));
    }

    return posts;
}

function benchmarkStreamingRanking(maxPosts = 1000000) {
    console.log(`🚀 JavaScript Streaming Ranking Engine Benchmark (up to ${maxPosts.toLocaleString()} posts)`);
    console.log("=" * 70);

    const results = {};
    const testScales = [10000, 100000, 500000, 1000000].filter(scale => scale <= maxPosts);

    for (const scale of testScales) {
        console.log(`\n📊 Testing ${scale.toLocaleString()} posts...`);
        results[scale] = {};

        const posts = generateTestPosts(scale);
        const engine = new StreamingRankingEngine(50000);

        const algorithms = ["hot_score", "engagement_score", "time_decay"];

        for (const algorithm of algorithms) {
            console.log(`  Running ${algorithm}...`);
            
            const result = engine.rankPostsFromList(posts, algorithm, 100);
            
            results[scale][algorithm] = {
                total_posts: result.totalPostsProcessed,
                processing_time_seconds: result.processingTimeSeconds,
                posts_per_second: result.totalPostsProcessed / result.processingTimeSeconds,
                memory_usage_mb: result.memoryUsageMb,
                memory_per_post_kb: (result.memoryUsageMb * 1024) / result.totalPostsProcessed,
                top_score: result.topPosts[0]?.[0] || 0,
                batch_size: result.batchSize
            };

            console.log(`    ✅ ${result.totalPostsProcessed.toLocaleString()} posts in ${result.processingTimeSeconds.toFixed(3)}s`);
            console.log(`    📈 ${(result.totalPostsProcessed / result.processingTimeSeconds).toFixed(0)} posts/sec`);
            console.log(`    💾 ${result.memoryUsageMb.toFixed(1)} MB memory`);
        }
    }

    return results;
}

function printBenchmarkResults(results) {
    console.log("\n" + "=" * 80);
    console.log("📊 JAVASCRIPT RANKING ENGINE BENCHMARK RESULTS");
    console.log("=" * 80);

    for (const [scale, algorithms] of Object.entries(results)) {
        console.log(`\n🎯 ${parseInt(scale).toLocaleString()} Posts:`);
        console.log("-" * 50);

        for (const [algorithm, metrics] of Object.entries(algorithms)) {
            console.log(`\n  ${algorithm.toUpperCase()}:`);
            console.log(`    ⏱️  Processing Time: ${metrics.processing_time_seconds.toFixed(3)}s`);
            console.log(`    🚀 Throughput: ${metrics.posts_per_second.toFixed(0)} posts/sec`);
            console.log(`    💾 Memory Usage: ${metrics.memory_usage_mb.toFixed(1)} MB`);
            console.log(`    📊 Memory per Post: ${metrics.memory_per_post_kb.toFixed(3)} KB`);
            console.log(`    🏆 Top Score: ${metrics.top_score.toFixed(6)}`);
        }
    }
}

// Export for use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        Post,
        StreamingRankingEngine,
        generateTestPosts,
        benchmarkStreamingRanking,
        printBenchmarkResults
    };
}

// Run benchmark if called directly
if (typeof require !== 'undefined' && require.main === module) {
    const results = benchmarkStreamingRanking(100000);
    printBenchmarkResults(results);
    
    // Save results to file
    const fs = require('fs');
    fs.writeFileSync('javascript_benchmark_results.json', JSON.stringify(results, null, 2));
    console.log("\n✅ Results saved to javascript_benchmark_results.json");
} 

File: ranking_engine.py
import time
import math
import heapq
from typing import Dict, List, Tuple, Optional, Any
from dataclasses import dataclass, field
from collections import defaultdict
import numpy as np


@dataclass
class Post:
    """Optimized post data structure for fast access and updates."""
    post_id: str
    likes: int = 0
    comments: int = 0
    shares: int = 0
    upvotes: int = 0
    downvotes: int = 0
    timestamp: float = field(default_factory=time.time)
    base_score: float = 0.0
    
    def __post_init__(self):
        """Calculate initial base score."""
        self.base_score = self.likes + self.comments * 2 + self.shares * 3


class RankingEngine:
    """
    High-performance ranking engine for social media posts.
    Designed to handle hundreds of thousands of posts efficiently.
    """
    
    def __init__(self, cache_size: int = 10000):
        self.posts: Dict[str, Post] = {}
        self.cache_size = cache_size
        self.ranking_cache: Dict[str, List[Tuple[float, str]]] = {}
        self.cache_timestamps: Dict[str, float] = {}
        self.cache_ttl = 60  # Cache for 60 seconds
        
    def add_post(self, post_id: str, **kwargs) -> None:
        """Add a new post to the ranking system."""
        if post_id in self.posts:
            raise ValueError(f"Post {post_id} already exists")
        
        self.posts[post_id] = Post(post_id=post_id, **kwargs)
        self._invalidate_cache()
        
    def update_post(self, post_id: str, **kwargs) -> None:
        """Update an existing post's metrics."""
        if post_id not in self.posts:
            raise ValueError(f"Post {post_id} not found")
        
        post = self.posts[post_id]
        for key, value in kwargs.items():
            if hasattr(post, key):
                setattr(post, key, value)
        
        # Recalculate base score
        post.base_score = post.likes + post.comments * 2 + post.shares * 3
        self._invalidate_cache()
        
    def _invalidate_cache(self) -> None:
        """Invalidate all cached rankings."""
        self.ranking_cache.clear()
        self.cache_timestamps.clear()
        
    def _is_cache_valid(self, algorithm: str) -> bool:
        """Check if cached ranking is still valid."""
        if algorithm not in self.cache_timestamps:
            return False
        return time.time() - self.cache_timestamps[algorithm] < self.cache_ttl
        
    def _hot_score(self, post: Post) -> float:
        """Calculate Reddit-style hot score."""
        order = math.log10(max(abs(post.upvotes - post.downvotes), 1))
        sign = 1 if post.upvotes - post.downvotes > 0 else -1 if post.upvotes - post.downvotes < 0 else 0
        seconds = time.time() - post.timestamp - 1134028003
        return round(sign * order + seconds / 45000, 7)
        
    def _engagement_score(self, post: Post) -> float:
        """Calculate engagement-based score."""
        total_engagement = post.likes + post.comments * 2 + post.shares * 3
        time_factor = time.time() - post.timestamp + 1
        return total_engagement / time_factor
        
    def _time_decay_score(self, post: Post, decay_rate: float = 0.01) -> float:
        """Calculate time-decay score."""
        time_elapsed = time.time() - post.timestamp
        return post.base_score * math.exp(-decay_rate * time_elapsed)
        
    def _hybrid_score(self, post: Post, weights: Dict[str, float] | None = None) -> float:
        """Calculate hybrid multi-factor score."""
        if weights is None:
            weights = {
                'likes': 1.0,
                'comments': 2.0,
                'shares': 3.0,
                'time_decay': 0.1
            }
        
        engagement = (post.likes * weights['likes'] + 
                     post.comments * weights['comments'] + 
                     post.shares * weights['shares'])
        
        time_elapsed = time.time() - post.timestamp
        time_factor = math.exp(-weights['time_decay'] * time_elapsed)
        
        return engagement * time_factor
        
    def _calculate_scores(self, algorithm: str, **kwargs) -> List[Tuple[float, str]]:
        """Calculate scores for all posts using the specified algorithm."""
        scores = []
        
        for post_id, post in self.posts.items():
            if algorithm == "hot_score":
                score = self._hot_score(post)
            elif algorithm == "engagement_score":
                score = self._engagement_score(post)
            elif algorithm == "time_decay":
                decay_rate = kwargs.get('decay_rate', 0.01)
                score = self._time_decay_score(post, decay_rate)
            elif algorithm == "hybrid":
                weights = kwargs.get('weights')
                score = self._hybrid_score(post, weights)
            else:
                raise ValueError(f"Unknown algorithm: {algorithm}")
                
            scores.append((score, post_id))
            
        return scores
        
    def get_ranked_posts(self, algorithm: str = "hot_score", limit: int = 100, 
                        **kwargs) -> List[Dict[str, Any]]:
        """
        Get ranked posts using the specified algorithm.
        Returns list of post dictionaries with ranking information.
        """
        # Check cache first
        if self._is_cache_valid(algorithm):
            cached_scores = self.ranking_cache[algorithm]
        else:
            # Calculate new scores
            scores = self._calculate_scores(algorithm, **kwargs)
            
            # Sort by score (descending) and limit
            scores.sort(reverse=True)
            
            # Cache the results
            self.ranking_cache[algorithm] = scores
            self.cache_timestamps[algorithm] = time.time()
            
            cached_scores = scores
            
        # Return top posts with ranking info
        result = []
        for rank, (score, post_id) in enumerate(cached_scores[:limit], 1):
            post = self.posts[post_id]
            result.append({
                'rank': rank,
                'post_id': post_id,
                'score': score,
                'likes': post.likes,
                'comments': post.comments,
                'shares': post.shares,
                'upvotes': post.upvotes,
                'downvotes': post.downvotes,
                'timestamp': post.timestamp,
                'age_hours': (time.time() - post.timestamp) / 3600
            })
            
        return result
        
    def get_post_rank(self, post_id: str, algorithm: str = "hot_score", **kwargs) -> Optional[int]:
        """Get the rank of a specific post."""
        ranked_posts = self.get_ranked_posts(algorithm, limit=len(self.posts), **kwargs)
        
        for post_data in ranked_posts:
            if post_data['post_id'] == post_id:
                return post_data['rank']
                
        return None
        
    def batch_add_posts(self, posts_data: List[Dict[str, Any]]) -> None:
        """Efficiently add multiple posts at once."""
        for post_data in posts_data:
            post_data_copy = post_data.copy()
            post_id = post_data_copy.pop('post_id')
            self.posts[post_id] = Post(post_id=post_id, **post_data_copy)
            
        self._invalidate_cache()
        
    def get_stats(self) -> Dict[str, Any]:
        """Get system statistics."""
        total_posts = len(self.posts)
        total_likes = sum(post.likes for post in self.posts.values())
        total_comments = sum(post.comments for post in self.posts.values())
        total_shares = sum(post.shares for post in self.posts.values())
        
        return {
            'total_posts': total_posts,
            'total_likes': total_likes,
            'total_comments': total_comments,
            'total_shares': total_shares,
            'avg_likes_per_post': total_likes / total_posts if total_posts > 0 else 0,
            'avg_comments_per_post': total_comments / total_posts if total_posts > 0 else 0,
            'avg_shares_per_post': total_shares / total_posts if total_posts > 0 else 0,
            'cache_size': len(self.ranking_cache),
            'memory_usage_mb': self._estimate_memory_usage()
        }
        
    def _estimate_memory_usage(self) -> float:
        """Estimate memory usage in MB."""
        # Rough estimation: each post object ~200 bytes
        post_memory = len(self.posts) * 200
        cache_memory = len(self.ranking_cache) * 1000  # Rough estimate for cache
        return (post_memory + cache_memory) / (1024 * 1024)  # Convert to MB


class OptimizedRankingEngine(RankingEngine):
    """
    Further optimized version using numpy for vectorized operations.
    Best for very large datasets (100k+ posts).
    """
    
    def __init__(self, cache_size: int = 10000):
        super().__init__(cache_size)
        self.post_arrays = {}  # Store post data as numpy arrays for vectorization
        
    def _vectorized_calculate_scores(self, algorithm: str, **kwargs) -> List[Tuple[float, str]]:
        """Vectorized score calculation using numpy for better performance."""
        if not self.posts:
            return []
            
        # Convert to numpy arrays for vectorized operations
        post_ids = list(self.posts.keys())
        likes = np.array([self.posts[pid].likes for pid in post_ids])
        comments = np.array([self.posts[pid].comments for pid in post_ids])
        shares = np.array([self.posts[pid].shares for pid in post_ids])
        upvotes = np.array([self.posts[pid].upvotes for pid in post_ids])
        downvotes = np.array([self.posts[pid].downvotes for pid in post_ids])
        timestamps = np.array([self.posts[pid].timestamp for pid in post_ids])
        
        current_time = time.time()
        time_elapsed = current_time - timestamps
        
        if algorithm == "hot_score":
            votes_diff = upvotes - downvotes
            order = np.log10(np.maximum(np.abs(votes_diff), 1))
            sign = np.sign(votes_diff)
            seconds = time_elapsed - 1134028003
            scores = sign * order + seconds / 45000
            
        elif algorithm == "engagement_score":
            engagement = likes + comments * 2 + shares * 3
            scores = engagement / (time_elapsed + 1)
            
        elif algorithm == "time_decay":
            decay_rate = kwargs.get('decay_rate', 0.1)
            base_scores = likes + comments * 2 + shares * 3
            scores = base_scores * np.exp(-decay_rate * time_elapsed)
            
        elif algorithm == "hybrid":
            weights = kwargs.get('weights', {
                'likes': 1.0, 'comments': 2.0, 'shares': 3.0, 'time_decay': 0.1
            })
            engagement = (likes * weights['likes'] + 
                         comments * weights['comments'] + 
                         shares * weights['shares'])
            time_factor = np.exp(-weights['time_decay'] * time_elapsed)
            scores = engagement * time_factor
            
        else:
            raise ValueError(f"Unknown algorithm: {algorithm}")
            
        # Create list of (score, post_id) tuples
        score_tuples = [(float(score), post_id) for score, post_id in zip(scores, post_ids)]
        return score_tuples 